File: fileinfo.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

# oyster - a python-based jukebox and web-frontend
#
# Copyright (C) 2004 Benjamin Hanzelmann <ben@nabcos.de>,
#  Stephan Windmüller <windy@white-hawk.de>,
#  Stefan Naujokat <git@ethric.de>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import taginfo
import urllib
import re
import cgitb
cgitb.enable()

import config
myconfig = config.get_config()
basedir = myconfig['basedir']
mediadir = myconfig['mediadir'][:-1]
import cgi
form = cgi.FieldStorage()

playlist = config.get_playlist()

import common
common.navigation_header()

try:
    soundfile = form['file'].value
except KeyError:
    soundfile = ''

import fifocontrol
if 'action' in form:
    fifocontrol.do_action(form['action'].value, soundfile)

import os.path
import sys
if not os.path.exists(mediadir + soundfile):
    print "<h1>Error!</h1>"
    print "<p>File <strong>" + soundfile + "</strong> could not be found.</p>"
    print "</body></html>"
    sys.exit()

if os.path.exists(myconfig['basedir']):
    oysterruns = 1
else:
    oysterruns = 0

print "<p><a href='browse.py?dir=/'>Mediadir</a>"

subdir = soundfile.replace(mediadir, '', 1)
subdir = os.path.dirname(subdir)
soundfileonly = os.path.basename(soundfile)
dirs = subdir.split('/')
incdir = ''
for partdir in dirs:
    escapeddir = urllib.quote(incdir + partdir)
    print "<a href='browse.py?dir=" + escapeddir + "'>" + partdir + "</a> / "
    incdir = incdir + partdir + "/"

print cgi.escape(soundfileonly) + "</p><br clear='all'/>"

isblacklisted = 0
if os.path.exists(myconfig['savedir'] + "blacklists/" + playlist):
    blacklist = open(myconfig['savedir'] + "blacklists/" + playlist)
    for rule in blacklist.readlines():
        if re.match('.*' + rule[:-1] + '.*', soundfile):
            isblacklisted = 1
    blacklist.close()

escapedfile = urllib.quote(soundfile)

if not os.access(mediadir + soundfile, os.R_OK):
    print "<h1>Sorry, Oyster does not have the permission to read this file!</h1>"
    print "</body></html>"
    sys.exit()

(mayVote, reason) = common.may_vote(soundfile, None)

print "<p>"
if oysterruns and mayVote:
    print "<span class='file'><a class='file' href='home.py?vote=" + escapedfile + "' >Vote this song</a></span>"
elif oysterruns and not mayVote:
    print "<span class='file' style='font-style: italic;' '>" + reason + "</span>"
else:
    print ""
print "</p>"

print "<p>"

if isblacklisted:
    print "<span class='blacklisted'>This song is blacklisted</span>"
else:
    regexfile = urllib.quote("^" + re.escape(soundfile) + "$")
    print "<a class='file' href='blacklist.py?affects=" + regexfile + "&amp;action=add'>Add this song to blacklist</a>"
print "</p>"

regexfile = urllib.quote("^" + re.escape(soundfile) + "$")

tag = taginfo.get_tag(mediadir + soundfile)

timesplayed = 0
logmatcher = re.compile('\A[0-9]{4}[0-9]{2}[0-9]{2}\-[0-9]{2}[0-9]{2}[0-9]{2} ([^ ]*) (.*)\Z')
log = open(myconfig['savedir'] + "logs/" + playlist)
for line in log.readlines():
    matcher = logmatcher.match(line[:-1])
    if matcher is not None and matcher.group(2).find(soundfile) > -1 and matcher.group(1) == 'DONE':
        timesplayed += 1

log.close()

albumdir = os.path.dirname(mediadir + soundfile) + "/"
coverdata = common.get_cover(albumdir, "200")

print "<table border='0'>"
if 'title' in tag:
    print "<tr><td class='fileinfo'>Title: </td><td>" + tag['title']

    if 'artist' in tag and 'title' in tag:
        print "<a class='file' href='lyrics.py?artist=" + urllib.quote(tag['artist']) + \
              "&amp;song=" + urllib.quote(tag['title']) + "'> (Songtext)</a>"

    print "</td></tr>"

if 'artist' in tag:
    print "<tr><td class='fileinfo'>Artist: </td><td>"
    print "<a href='search.py?searchtype=normal&amp;playlist=current&amp;" + \
        "search=" + urllib.quote(tag['artist']) + "' title='Search for " + \
        "this artist' class='file'>" + tag['artist'] + "</a></td></tr>"

if coverdata != '':
    print "<tr><td class='fileinfo'>Cover: </td><td>" + coverdata + "</td></tr>"

tagtuple = (
    ('Album: ', 'album'),
    ('Track: ', 'track'),
    ('Year: ', 'year'),
    ('Genre: ', 'genre'),
    ('Comment: ', 'comment'),
    ('Format: ', 'format'),
    ('Length: ', 'playtime')
)

for line in tagtuple:
    if line[1] in tag:
        print "<tr><td class='fileinfo'>" + line[0] + "</td>" + \
            "<td>" + tag[line[1]] + "</td></tr>"

print "<tr><td colspan='2'>&nbsp;</td></tr>"
print "<tr><td class='fileinfo'>Played: </td><td>" + str(timesplayed) + " times</td></tr>"
print "<tr><td class='fileinfo'>Score: </td>"
print "<td><a href='fileinfo.py?action=scoredown&amp;file=" + escapedfile + "' title='Score down'>"
print "<img src='themes/" + myconfig['theme'] + "/scoredownfile.png' border='0' alt='-'/></a> "
print "<strong>" + str(tag['score']) + "</strong>"
print " <a href='fileinfo.py?action=scoreup&amp;file=" + escapedfile + "' title='Score up'>"
print "<img src='themes/" + myconfig['theme'] + "/scoreupfile.png' border='0' alt='+'/></a></td></tr>"

print "</table>"

print "</body></html>"

File: fileinfo.pl
#!/usr/bin/perl
use CGI qw/:standard -no_xhtml/;
use URI::Escape;
use strict;
use oyster::conf;
use oyster::taginfo;
use oyster::fifocontrol;
use oyster::common;

my %config = oyster::conf->get_config('oyster.conf');
my $playlist = oyster::conf->get_playlist();

oyster::common->navigation_header();

my $mediadir = $config{'mediadir'};
$mediadir =~ s/\/$//;
my $file = param('file') || '';

if (param('action')) {
    oyster::fifocontrol->do_action(param('action'), $file, '');
}   

if (!(-e "$mediadir$file")) {
    print h1('Error!');
    print "<p>File <strong>$file</strong> could not be found.</p>";
    print end_html;
    exit 0;
}

my $oysterruns = 0;

if (-e $config{'basedir'}) {
    $oysterruns = 1;
}

print "<p>Info for ";

my $subdir = my $fileonly = $file;
$subdir =~ s/^\Q$mediadir\E//;
$subdir =~ s/\/[^\/]*$//;
$fileonly =~ s/^.*\///;
my @dirs = split(/\//, $subdir);
my $incdir = '';
foreach my $partdir (@dirs) {
    my $escapeddir = uri_escape("$incdir$partdir", "^A-Za-z");
    print "<a href='browse.pl?dir=$escapeddir'>$partdir</a> / ";
    $incdir = $incdir . "$partdir/";
}

print "$fileonly</p><br clear='all'>\n";

my $isblacklisted = 0;
my $playlist = oyster::conf->get_playlist();
open (BLACKLIST, "${config{'savedir'}}blacklists/$playlist");
while (my $rule = <BLACKLIST>) {
    chomp($rule);
    $isblacklisted = 1 if ($file =~ /$rule/);
}
close (BLACKLIST);

my $escapedfile = uri_escape("$file", "^A-Za-z");

print "<table width='100%'><tr>";
if ($oysterruns) {
    print "<td align='left'><a class='file' href='oyster-gui.pl?vote=$escapedfile' target='curplay'>Vote for this song</a></td>\n";
} else {
    print "<td></td>\n";
}
my $regexpfile = uri_escape("^$file\$", "^A-Za-z");

if ($isblacklisted) {
    print "<td align='right'><span class='blacklisted'>File is blacklisted</span></td></tr></table>";
} else {
    print "<td align='right'><a class='file' href='blacklist.pl?affects=${regexpfile}&amp;action=add'>Add this song to Blacklist</td></tr></table>";
}

my %tag = oyster::taginfo->get_tag("$mediadir$file");

my $timesplayed = 0;
open (LOG, "${config{'savedir'}}logs/$playlist");
while (my $line = <LOG>) {
    my ($year, $month, $day, $hour, $minute, $second, $playreason, $filename);
    chomp($line);
    $_ = $line;
    ($year, $month, $day, $hour, $minute, $second, $playreason, $filename) =
	m@^([0-9]{4})([0-9]{2})([0-9]{2})\-([0-9]{2})([0-9]{2})([0-9]{2})\ ([^\ ]*)\ (.*)$@;
    if ($filename =~ /\Q$file\E/) {
	if ($playreason eq 'DONE') {
	    $timesplayed++;
	}
    }
}
close LOG;

my $albumdir = $mediadir . $file;
$albumdir =~ s/[^\/]*$//;
my $coverdata = oyster::common->get_cover($albumdir);

print "<table cellpadding='10'>";
if ($tag{'title'}) {
    print "<tr><td><strong>Title</strong></td><td>$tag{'title'}</td><td rowspan='6'>$coverdata</td></tr>";
} else {
    print "<tr><td><strong>Title</strong></td><td>$tag{'title'}</td></tr>";
}
print "<tr><td><strong>Artist</strong></td><td>$tag{'artist'}</td></tr>" if ($tag{'artist'});
print "<tr><td><strong>Album</strong></td><td>$tag{'album'}</td></tr>" if ($tag{'album'});
print "<tr><td><strong>Track Number</strong></td><td>$tag{'track'}</td></tr>" if ($tag{'track'});
print "<tr><td><strong>Year</strong></td><td>$tag{'year'}</td></tr>" if ($tag{'year'});
print "<tr><td><strong>Genre</strong></td><td>$tag{'genre'}</td></tr>" if ($tag{'genre'});
print "<tr><td><strong>Comment</strong></td><td>$tag{'comment'}</td></tr>" if ($tag{'comment'});
print "<tr><td><strong>File Format</strong></td><td>$tag{'format'}</td></tr>";
print "<tr><td><strong>Playtime</strong></td><td>$tag{'playtime'}</td></tr>" if ($tag{'playtime'});
print "<tr><td colspan='2'>&nbsp;</td></tr>";
print "<tr><td><strong>Times played</strong></td><td>$timesplayed</td></tr>";
print "<tr><td><strong>Current Oyster-Score</strong></td>";
print "<td><a href='fileinfo.pl?action=scoredown&file=$escapedfile'><img src='themes/${config{'theme'}}/scoredownfile.png' border='0' alt='-'></a> ";
print "<strong>$tag{'score'}</strong>";
print " <a href='fileinfo.pl?action=scoreup&file=$escapedfile'><img src='themes/${config{'theme'}}/scoreupfile.png' border='0' alt='+'></a></td></tr>";
print "</table>";

print end_html;
